### PR TITLE
Go: proper support for json.Number values

### PR DESCRIPTION
**Pull request by akireev-booking-com**
When unmarshalling free-form json in Go, numbers become either float64 or json.Number. With the first option you lose precision for large integers, so it makes sense to always use the second.

But in current implementation of sereal/encode.go the json.Number values are serialized as strings. So, if you do a conversion like this:

json -> data -> sereal -> data -> json,

all numbers in your structure will become strings, which is not good.

Since encoding/json is a core module, and interoperability with it is a practical issue, it makes sense to support proper serialization of json.Number right here, which this commit does.

### CODE REVIEW

File: Go/sereal/encode.go [hunk 1]
```diff
@@ -330,6 +335,26 @@ func (e *Encoder) encodeDouble(by []byte, f float64) []byte {
...
 
+func (e *Encoder) encodeJsonNumber(by []byte, n json.Number, isKeyOrClass bool, strTable map[string]int) []byte {
+	if int64Value, err := n.Int64(); err == nil {
+		return e.encodeInt(by, reflect.Int, int64Value)
+	}
+
+	// we do not want to lose precision for large integers, as those are often IDs of things
+	// so if a value looks like a positive integer, but does not fit to int64, better keep it as a string
+	// uint64 values which don't fit to signed int are a common example
+	bigInt, isBigInt := new(big.Int).SetString(n.String(), 10)
```
Comment (mbarbon) [inline Go/sereal/encode.go]: Maybe only try this branch if the previous error is `strconv.ErrRange`?

Or even better, if the previous error is `ErrRange`, just re-encode it straight to string?

Comment (akireev-booking-com) [reply Go/sereal/encode.go]: Yes, make sense to check for `strconv.ErrRange`. 

Also, on the second thought, stringifying large positive integers, but leaving large negative ones as floats with lost precision, goes against the principle of the least surprise, so yeah, we don't have to actually parse the number.

Updated the MR.

File: Go/sereal/encode.go [hunk 2]
```diff
@@ -330,6 +335,29 @@ func (e *Encoder) encodeDouble(by []byte, f float64) []byte {
 	return by
 }
 
+func (e *Encoder) encodeJsonNumber(by []byte, n json.Number, isKeyOrClass bool, strTable map[string]int) []byte {
+	int64Value, err := n.Int64()
+	if err == nil {
+		return e.encodeInt(by, reflect.Int, int64Value)
+	} else {
```
Comment (dgryski) [inline Go/sereal/encode.go:342]: This `else` is unnessary -- the `if` block returns.

File: Go/sereal/encode.go [hunk 3]
```diff
@@ -330,6 +335,29 @@ func (e *Encoder) encodeDouble(by []byte, f float64) []byte {
...
+
+		// we do not want to lose precision for large integers, as those are often IDs or hashsums of things
+		// so if the error we got is about the number not fitting into the int64 datatype, better keep it as a string instead of converting to float and losing precision
+		// uint64 values which don't fit to signed int are a common example, but the value might be of arbitrary size
+
+		numErr, isNumErr := err.(*strconv.NumError)
```
Comment (dgryski) [inline Go/sereal/encode.go:348]: `isNumErr` here usually called `ok` for type assertions.

**Pull request merged**